File: packages/ts/src/strategies/remove-comments.ts
/**
 * remove-comments: Strip // line comments, /* block comments *\/, and # line comments
 * from JSON-like text. String-aware: does not strip inside quoted strings.
 * Runs before structural strategies.
 */
export const removeComments = (text: string): string => {
  let result = "";
  let i = 0;
  let inString = false;
  let escaped = false;

  while (i < text.length) {
    const ch = text[i];
    if (escaped) {
      result += ch;
      escaped = false;
      i++;
      continue;
    }
    if (inString) {
      if (ch === "\\") { escaped = true; result += ch; i++; continue; }
      if (ch === '"') inString = false;
      result += ch;
      i++;
      continue;
    }
    // Not in string
    if (ch === '"') { inString = true; result += ch; i++; continue; }
    // // comment
    if (ch === "/" && text[i + 1] === "/") {
      while (i < text.length && text[i] !== "\n") i++;
      continue;
    }
    // /* comment */
    if (ch === "/" && text[i + 1] === "*") {
      i += 2;
      while (i < text.length && !(text[i] === "*" && text[i + 1] === "/")) i++;
      i += 2;
      continue;
    }
    // # comment (only at start of token, not inside strings)
    if (ch === "#") {
      while (i < text.length && text[i] !== "\n") i++;
      continue;
    }
    result += ch;
    i++;
  }
  return result;
};

File: packages/python/src/output_guard/strategies/remove_comments.py
"""remove-comments: Strip //, /* */ and # comments. String-aware.

Mirror of packages/ts/src/strategies/remove-comments.ts.
"""

from __future__ import annotations


def remove_comments(text: str) -> str:
    out: list[str] = []
    i = 0
    n = len(text)
    in_string = False
    escaped = False
    while i < n:
        ch = text[i]
        if escaped:
            out.append(ch)
            escaped = False
            i += 1
            continue
        if in_string:
            if ch == "\\":
                escaped = True
                out.append(ch)
                i += 1
                continue
            if ch == '"':
                in_string = False
            out.append(ch)
            i += 1
            continue
        if ch == '"':
            in_string = True
            out.append(ch)
            i += 1
            continue
        if ch == "/" and i + 1 < n and text[i + 1] == "/":
            while i < n and text[i] != "\n":
                i += 1
            continue
        if ch == "/" and i + 1 < n and text[i + 1] == "*":
            i += 2
            while i < n and not (text[i] == "*" and i + 1 < n and text[i + 1] == "/"):
                i += 1
            i += 2
            continue
        if ch == "#":
            while i < n and text[i] != "\n":
                i += 1
            continue
        out.append(ch)
        i += 1
    return "".join(out)
